### PR TITLE
feat: (levm) Ethereum Foundation Tests

### LINKED ISSUE #580: Ethereum Foundation Tests
Run all the `GeneralStateTests` of the EF tests located [here](https://github.com/ethereum/tests/tree/develop/GeneralStateTests).

The EVM Mlir project already has all the scaffolding code necessary to run them [here](https://github.com/lambdaclass/evm_mlir/tree/main/tests/ef_tests_executor), we can use that (I suggest to start by running the tests and looking at [this function](https://github.com/lambdaclass/evm_mlir/blob/main/tests/ef_tests.rs#L81))

For now we only target `Cancun`

### PR DESCRIPTION
**Pull request by iinaki**
**Summary**
Implemented a way to run the Ethereum Foundation tests for the LEVM.

**Changes**
- Setup the `TxEnv`, `BlockEnv` and `Db` from a unique test to initialize a new VM for that test.
- Created the `vm.transact()` method, which executes the VM and obtains the execution result.
- Updated the `ignored_groups` from the EF tests to check which tests pass and which are ignored.
- Currently we have 120 EF tests passing from 2807, the rest are skipped.

Closes #580

### CODE REVIEW

**Comment by jrchatruc** [comment]
The [tests_v3.0.0.tar.gz](https://github.com/lambdaclass/ethereum_rust/pull/600/files#diff-9cb2ce89489590eaabc8eab36c73a7c200fdcf26bc79e3071e2a3447b1de7d83) file should not be commited, we can add it to the `gitignore` file after removing it as well

**Comment by iinaki** [comment]
> The [tests_v3.0.0.tar.gz](https://github.com/lambdaclass/ethereum_rust/pull/600/files#diff-9cb2ce89489590eaabc8eab36c73a7c200fdcf26bc79e3071e2a3447b1de7d83) file should not be commited, we can add it to the `gitignore` file after removing it as well

Done!

File: tests_v3.0.0.tar.gz
[the remark this answers is not in the document]
Comment (iinaki) [reply tests_v3.0.0.tar.gz:1]: nope, removed!

File: crates/vm/levm/src/vm.rs [hunk 1]
```diff
@@ -216,19 +276,94 @@ impl VM {
...
+        // let gas_remaining = initial_gas - gas_used;
+
+        // TODO: Probably here we need to add the access_list_cost to gas_used, but we need a refactor of most tests
+        let gas_refunded = self.env.refunded_gas.min(gas_used / GAS_REFUND_DENOMINATOR);
+
+        // let exit_status = res
+        //     .inner_context
+        //     .exit_status
+        //     .clone()
+        //     .unwrap_or(ExitStatusCode::Default);
```
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/vm.rs]: done!

File: crates/vm/levm/src/vm.rs [hunk 2]
```diff
@@ -216,19 +276,94 @@ impl VM {
...
-    pub fn execute(&mut self) -> ExecutionResult {
-        let mut current_call_frame = self
-            .call_frames
-            .pop()
-            .expect("Fatal Error. This shouldn't happen"); // if this happens during execution, we are cooked 💀
+    pub fn get_result(&mut self, res: ExecutionResult) -> Result<ResultAndState, VMError> {
+        // let initial_gas = self.env.gas_limit;
+        let gas_used = self.env.consumed_gas;
+
+        // let gas_remaining = initial_gas - gas_used;
```
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/vm.rs]: done!

File: crates/vm/levm/src/vm.rs [hunk 3]
```diff
@@ -165,11 +227,8 @@ fn address_to_word(address: Address) -> U256 {
     U256::from_str(&format!("{address:?}")).unwrap()
 }
 
-// The execution model specifies how the system state is
-// altered given a series of bytecode instructions and a small
-// tuple of environmental data.
-
 impl VM {
+    // TODO: block and transaction, not this
```
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/vm.rs]: done!

File: crates/vm/levm/src/vm.rs [hunk 4]
```diff
@@ -124,6 +141,52 @@ impl Db {
...
+
+    pub fn into_state(&self) -> HashMap<Address, StateAccount> {
+        self.accounts
+            .iter()
+            .map(|(address, acc)| {
+                // let code = acc
+                //     .has_code()
+                //     .then_some(&acc.bytecode_hash())
+                //     .into_iter().flatten()
+                //     .cloned();
```
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/vm.rs]: done!

File: crates/vm/levm/src/vm.rs [hunk 5]
```diff
@@ -3,33 +3,40 @@ use std::{
...
+    },
 };
 extern crate ethereum_rust_rlp;
 use ethereum_rust_rlp::encode::RLPEncode;
 use ethereum_types::H160;
 use sha3::{Digest, Keccak256};
 
-#[derive(Clone, Default, Debug)]
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+// TODO: complete account abstraction
```
[the remark this answers is not in the document]
Comment (iinaki) [reply crates/vm/levm/src/vm.rs]: done!

**Review by lazcanoluca — APPROVED** [review]
Great job! 🚀

**Review by jrchatruc — APPROVED** [review]

**Pull request merged**